File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    stack<int> stack,minstack;
    MinStack() {
        // stack - [-2,0,-3]
        // minstack - [-2,-3]
    }
    
    void push(int val) {
        // push into stack , push to minstack only if val<minstack.top();
        stack.push(val);
        if(minstack.empty()){
            minstack.push(val);
        }
        else if(minstack.size() && minstack.top()>=val){
            minstack.push(val);
        }
    }
    
    void pop() {
        int top_val=stack.top();
        if(minstack.top()==top_val){
            minstack.pop();
        }
        stack.pop();
    }
    
    int top() {
        return stack.top();
    }
    
    int getMin() {
        return minstack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.st = []
        self.minst = []

    def push(self, val: int) -> None:
        self.st.append(val) 
        if not(self.minst) or self.minst[-1]>=val:
            self.minst.append(val)

    def pop(self) -> None:
        top = self.st.pop()
        if self.minst[-1] == top: 
            self.minst.pop()
        
    def top(self) -> int:
        return self.st[-1]

    def getMin(self) -> int:
        return self.minst[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()